Student/Student/Courses/Syllabus/SyllabusActionableItemsPresenter.swift

```diff
@@ -0,0 +1,137 @@
+        }
+    }()
+
+    init(env: AppEnvironment = .shared, view: SyllabusActionableItemsViewProtocol, courseID: String, sort: GetAssignments.Sort = .position) {
+        self.courseID = courseID
+        self.env = env
+        self.view = view
+        self.sort = sort
+    }
+
+    func viewIsReady() {
+        assignments.exhaust(while: { _ in true })
+        calendarEvents.exhaust(while: { _ in true })
+        course.refresh()
+        update()
+    }
+
+    func update() {
+        var models = assignmentViewModels + calendarViewModels
+        models.sort { (a, b) -> Bool in
+            return (a.dueDate ?? .distantFuture).compare(b.dueDate ?? .distantFuture) == .orderedAscending
+        }
+        view?.update(models: models)
+        loadColor()
+    }
+
+    func updateAssignments() {
+        assignmentUpdateCount += 1
+        assignmentViewModels = assignments.map {
+            SyllabusActionableItemsViewController.ViewModel(id: $0.id, htmlUrl: $0.htmlURL, title: $0.name, dueDate: $0.dueAt, formattedDate: formattedDueDate($0.dueAt), image: icon(for: $0))
+        }
+        signalUpdate()
+    }
+
+    func updateCalendarEvents() {
+        calendarEventUpdateCount += 1
+        calendarViewModels = calendarEvents.map {
+            SyllabusActionableItemsViewController.ViewModel(id: $0.id, htmlUrl: $0.routingURL,
+                                                           title: $0.title, dueDate: $0.endAt, formattedDate: formattedDueDate($0.endAt), image: .icon(.calendarMonth, .line))
+        }
+        signalUpdate()
+    }
+
+    func signalUpdate() {
+        let updateThreshold = 2
+        if(assignmentUpdateCount >= updateThreshold && calendarEventUpdateCount >= updateThreshold) {
+            update()
+        }
+    }
+
+    func loadColor() {
+        guard let course = course.first else { return }
+        view?.updateNavBar(subtitle: course.name, color: course.color)
+    }
+
+    func select(_ htmlURL: URL, from: UIViewController) {
+        env.router.route(to: htmlURL, from: from, options: nil)
+    }
+
+    func formattedDueDate(_ date: Date?) -> String {
+        var result = NSLocalizedString("No Due Date", comment: "")
+        if let date = date {
+            result = DateFormatter.localizedString(from: date, dateStyle: .medium, timeStyle: .short)
+        }
+        return result
+    }
+
+    func icon(for: Assignment) -> UIImage? {
+        let assignment = `for`
+        var image: UIImage? = .icon(.assignment, .line)
+        if assignment.quizID != nil {
+            image = .icon(.quiz, .line)
+        } else if assignment.submissionTypes.contains(.discussion_topic) {
+            image = .icon(.discussion, .line)
+        } else if assignment.submissionTypes.contains(.external_tool) || assignment.submissionTypes.contains(.basic_lti_launch) {
+            image = .icon(.lti, .line)
+        }
+
+        if assignment.lockedForUser {
+            image = .icon(.lock, .line)
```

Comment:
Nice job handling locked state. Should we be disabling these? What does old syllabus do?

Reply:
locked assignment still goes to locked assignment.